### ISSUE TITLE
Windows, unable to run telnet

### ISSUE THREAD

**Issue by pehruby** [CiscoDevNet/virlutils#45]
I use 64 bit windows and 32 bit Python. virlutils is unable to run telnet client located in c:\windows\system32 due to Windows File System Redirector. See http://code.activestate.com/recipes/578035-disable-file-system-redirector/

Do you think there is a chance to adjust virlutils code so that such environment is taken into account?

**Comment by kecorbin** [comment]
@pehruby I'm open to anything - I don't have a windows dev envirornment at hand, any chance you have a PR of what is needed?

**Comment by pehruby** [comment]
@kecorbin Another problem is how to run "ssh -t" from Windows environment ... I think the better solution is to use Vagrant and Ubuntu :)

**Comment by kecorbin** [comment]
@pehruby no argument from me on that - but let me know, happy to enhance the windows experience, just need help testing ;)

**Comment by kecorbin** [comment]
@hpreston has done some windows testing - maybe he has some ideas.  Let's leave issue open for now as it's a valid one

**Comment by pehruby** [comment]
OK. Thanks. If you think I can help with testing, please let me know.

**Comment by hpreston** [comment]
@pehruby  I don't have my Windows test environment handy right now to verify, but a couple of thoughts.. 

I have moved to always/only using a bash style shell on Windows rather than CMD or PowerShell interfaces because I find that it results in a far more consistent experience across platforms.  This typically has me using `git-bash` for most work, and occasionally the full bash interface provided by the Linux Subsystem included with Windows 10.  

I know that sometime around W7, Microsoft stopped including `telnet.exe` in the operating system by default, and required installing it manually... have you installed it?  

I think the `virl ssh` command will work fine within `git-bash` or other bash options.  

And as for Vagrant + Ubuntu, I'm a huge fan of Vagrant for dev and testing, however I also find that if I want a topology larger than 2 or 3 nodes, and having a functioning data plane is important to what I'm working on, VIRL is a better choice.  Both by offloading the running of the VMs to a dedicated machine, but also ruling out oddities in the network stack used by VirtualBox with the VMs.  Also, the 4 vNic limit on VirtualBox can quickly become a limiting factor.  

hope this helps.

**Comment by pehruby** [comment]
@hpreston git-bash looks like a good choice. At least ssh works fine in this shell. But the problem with telnet persists. Yes, I've installed telnet manually. The program is installed in c:\windows\system32 and this causes the problem described above.

```text
$ virl console iosv-1
Attempting to connect to console of iosv-1
Exception raised while running your command
Please open an issue and provide this info:
[Error 2] Syst▒m nem▒▒e nal▒zt uveden▒ soubor
(venv)
```
[Error 2] The system cannot find given file


Look at this:
```text
(venv)
PeHruby@CZCL01506 MINGW64 /c/Users/pehruby/Documents/virlutils/cpoj-stage
$ python
Python 3.6.2 (v3.6.2:5fd33b5, Jul  8 2017, 04:14:34) [MSC v.1900 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> from subprocess import call
>>> call('cmd')
Microsoft Windows [Version 10.0.14393]
(c) 2016 Microsoft Corporation. Všechna práva vyhrazena.


C:\Users\pehruby\Documents\virlutils\cpoj-stage>dir c:\windows\system32\tel*
 Volume in drive C is OSDisk
 Volume Serial Number is B687-55DA

 Directory of c:\windows\system32

16.07.2016  13:43           107 520 telephon.cpl
               1 File(s)        107 520 bytes
               0 Dir(s)  43 488 952 320 bytes free
C:\Users\pehruby\Documents\virlutils\cpoj-stage>exit
0
>>> quit()
(venv)
PeHruby@CZCL01506 MINGW64 /c/Users/pehruby/Documents/virlutils/cpoj-stage
$ dir c:/windows/system32/tel*
c:/windows/system32/telephon.cpl  c:/windows/system32/telnet.exe
$ dir c:/windows/SysWOW64/tel*
c:/windows/SysWOW64/telephon.cpl
(venv)

```
As you can see telnet.exe is visible in c:/windows/system32/ but when I try to access it from cmd running from python, telnet is not visible because in fact c:\windows\SysWOW64 is used instead of system32.

Regarding Vagrant + Ubuntu. I'm using ubuntu as an environment for virlutils. VIRL is running on esxi.

**Comment by kecorbin** [comment]
@pehruby @hpreston I through this together in a branch https://github.com/kecorbin/virlutils/tree/45_windows_telnet it only addresses virl console but can be expanded to the others if it works. Let me know if you have a chance to install and test.

**Comment by pehruby** [comment]
Hi Kevin, thanks for fast response. I've tested your patch and it works fine from windows cmd prompt. But it doesn't work from git-bash.
```text
PeHruby@CZCL01506 MINGW64 /c/Users/pehruby/Documents/Programovani/virtualenv3/virlutils/2-ios-router (45_windows_telnet)
$ virl console router1
Attempting to connect to console of router1
(venv)
PeHruby@CZCL01506 MINGW64 /c/Users/pehruby/Documents/Programovani/virtualenv3/virlutils/2-ios-router (45_windows_telnet)
$ telnet 10.105.1.1
(venv)
PeHruby@CZCL01506 MINGW64 /c/Users/pehruby/Documents/Programovani/virtualenv3/virlutils/2-ios-router (45_windows_telnet)
$
```
Is it possible to detect environment which is virlutils running in and if Windows + git-bash is detected run telnet using winpty or start command ( 'winpty telnet 10.105.1.1 23')?

**Comment by kecorbin** [comment]
Yeah, I think for telnet and ssh we could have virlrc or ENV vars to allow the user to specify what command line should look like.

This would be extensible for many different use cases/operating systems

Sent from my iPhone

On Jul 15, 2018, at 4:44 AM, Petr Hruby <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi Kevin, thanks for fast response. I've tested your patch and it works fine from windows cmd prompt. But it doesn't work from git-bash.

PeHruby@CZCL01506 MINGW64 /c/Users/pehruby/Documents/Programovani/virtualenv3/virlutils/2-ios-router (45_windows_telnet)
$ virl console router1
Attempting to connect to console of router1
(venv)
PeHruby@CZCL01506 MINGW64 /c/Users/pehruby/Documents/Programovani/virtualenv3/virlutils/2-ios-router (45_windows_telnet)
$ telnet 10.105.1.1
(venv)
PeHruby@CZCL01506 MINGW64 /c/Users/pehruby/Documents/Programovani/virtualenv3/virlutils/2-ios-router (45_windows_telnet)
$


Is it possible to detect environment which is virlutils running in and if Windows + git-bash is detected run telnet using winpty or start command ( 'winpty telnet 10.105.1.1 23')?

**Comment by hpreston** [comment]
I like the idea of allowing optional virlrc entries to override the paths to "supporting programs" like the SSH and telnet clients.  Then as we find them, we can document examples for different programs...

**Comment by kecorbin** [comment]
@hpreston @pehruby just opened the PR, take a peek at the new configuration options I'm proposing and let me know if they make sense and work for your use cases. https://github.com/kecorbin/virlutils/tree/45_windows_telnet#other-configuration-options

**Comment by kecorbin** [comment]
Yes - I thought about including such in the current PR, but it was a little out of scope for the current issue.  We should track as separate issue and part of a broader cleanup of the current ability to proxy via management lxc IMO.

**Comment by pehruby** [comment]
@kecorbin OK.

**Comment by kecorbin** [comment]
closed #46
